qp_shogun/sortmerna/utils.py

```diff
@@ -0,0 +1,177 @@
+# The full license is in the file LICENSE, distributed with this software.
+# -----------------------------------------------------------------------------
+
+# -----------------------------------------------------------------------------
+# This file contains functions used by multiple commands
+# -----------------------------------------------------------------------------
+from qiita_client.util import system_call, get_sample_names_by_run_prefix
+from itertools import zip_longest
+from os.path import basename, join, exists
+from functools import partial
+from qiita_client import ArtifactInfo
+
+
+def make_read_pairs_per_sample(forward_seqs, reverse_seqs, map_file):
+    """Recovers read pairing information
+
+    Parameters
+    ----------
+    forward_seqs : list of strs
+        The list of forward seqs filepaths
+    reverse_seqs : list of str
+        The list of reverse seqs filepaths
+    map_file : str
+        The path to the mapping file
+
+    Returns
+    -------
+    samples: list of tup
+        list of 4-tuples with run prefix, sample name, fwd read fp, rev read fp
+
+    Raises
+    ------
+    ValueError
+        If the rev is not an empty list and the same length than fwd seqs
+        The prefixes of the run_prefix don't match the file names
+
+    Notes
+    -----
+    At this stage it is required that if reverse sequences are present that all
+    samples have both a forward and a reverse sequence. However, the read
+    trimming/filtering step can sometimes eliminate all reverse reads,
+    especially in low
+    coverage samples with poor overall reverse read quality.
+    """
+
+    # sort forward seqs
+    forward_seqs.sort()
+
+    # check that rev seqs are same len
+    if reverse_seqs:
+        if len(forward_seqs) != len(reverse_seqs):
+            raise ValueError('Your reverse and forward files are of different '
+                             'length. Forward: %s. Reverse: %s.' %
+                             (', '.join(forward_seqs),
+                              ', '.join(reverse_seqs)))
+        reverse_seqs.sort()
+
+    # get run prefixes
+    # These are prefixes that should match uniquely to forward reads
+    # sn_by_rp is dict of samples keyed by run prefixes
+    sn_by_rp = get_sample_names_by_run_prefix(map_file)
+
+    # make pairings
+    samples = []
+    used_prefixes = set()
+    for i, (fwd_fp, rev_fp) in enumerate(zip_longest(forward_seqs,
+                                                     reverse_seqs)):
+        # fwd_fp is the fwd read filepath
+        fwd_fn = basename(fwd_fp)
+
+        # iterate over run prefixes and make sure only one matches
+        run_prefix = None
+        for rp in sn_by_rp:
+            if fwd_fn.startswith(rp) and run_prefix is None:
+                run_prefix = rp
+            elif fwd_fn.startswith(rp) and run_prefix is not None:
+                raise ValueError('Multiple run prefixes match this fwd read: '
+                                 '%s' % fwd_fn)
```

Comment:
I would suggest also mentioning what the problematic run prefixes are.
```suggestion
                raise ValueError('Multiple run prefixes match this fwd read: '
                                 '%s. Matches are: %s and %s' % (fwd_fn, rp, run_prefix))
```